software/spmd/bsg_attr_noalias_ex/main.cpp

```diff
@@ -0,0 +1,320 @@
+// This kernel performs saxpy to demonstrate how the placement of
+// bsg_attr_noalias affects code generation.
+
+// NOTE: 
+
+// bsg_attr_noalias is NOT FULLY SUPPORTED IN LLVM. It works for 1D array
+// accesses, but your mileage may vary with anything more complex.
+
+
+// Our goal is to optimize saxpy (Scalar * X Plus Y) by unrolling it
+// by a factor of 4 to increase performance. Performance gains can be
+// attributed to two effects: 
+
+// 1) Increasing the distance between load and use instructions to
+// hide DRAM latency 
+
+// 2) Reduce loop control overhead.
+
+// In this example we will use the restrict type qualifier
+// (https://en.cppreference.com/w/c/language/restrict) on pointers to
+// indicate that unique pointer declarations represent non-overlapping
+// regions of memory (i.e. they do not alias).
+
+// restrict allows the compiler to reorder loads and stores to unique
+// pointer declarations because stores to a restricted pointer do not
+// modify the data of restricted pointer declarations.
+
+// Reordering means loads can be moved ahead of the stores of a
+// previous iteration to increase the instruction distance between
+// load and use.
+
+// Using bsg_attr_noalias is one of TWO critical steps to utilizing
+// non-blocking loads effectively in the manycore architecture.
+
+// For the second critical step see kernel.cpp in the remote
+// directory.
+
+// For a synthesis of the two, see remote+restrict
+
+
+
+// The C implementation of saxpy is in saxpy-c.c
+// The C++ implementation of saxpy is in saxpy-cpp.cpp
+
+// These sources are compiled independently from each other, and
+// main.c. They are linked together to form main.riscv. In this way we
+// can demonstrate/test the differences in C/C++.
+
+#include <bsg_manycore.h>
+#include <bsg_set_tile_x_y.h>
+// saxpy.h is generated by saxpy.py. It defines MACRO_A, MACRO_B, MACRO_GOLD, and N_ELS
+#include "saxpy.h"
+// saxpy-c.h and saxy-cpp.hpp contain the C and C++ function
+// definitions, respectively.
+#include "saxpy-c.h"
+#include "saxpy-cpp.hpp"
+#include <string.h>
+
+#define SAXPY_TAG_BASE 0
```

Comment:
Are these solely intended for profiling? If so, a small comment or macro reflecting that could be useful... these seem to need some time to decode.